### PR TITLE
[studio-ui] Guest page builder move mode

### PR DESCRIPTION
**Pull request by aachavarria**

### CODE REVIEW

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 1]
```diff
@@ -15,29 +15,76 @@
...
+  }, [index]);
+
+  const dropzoneRecord = useMemo(
+    () => fromElement(getParentElementFromICEProps(modelId, fieldId, index)[0] as Element),
+    [fieldId, index, modelId]
+  );
+
+  const dropzoneChildrenLength = useMemo(
+    () => compileDropZone(dropzoneRecord.iceIds[0]).children.length,
+    [dropzoneRecord.iceIds]
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]:
```suggestion
    [dropzoneRecord.iceIds[0]]
```

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 2]
```diff
@@ -56,16 +104,20 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
-        </UltraStyledIconButton>
-      </Tooltip>
+      {index !== 0 && (
+        <Tooltip title="Move up/left (← or ↑)">
+          <UltraStyledIconButton size="small" onClick={onMoveUp}>
+            <ArrowUpwardRoundedIcon />
+          </UltraStyledIconButton>
+        </Tooltip>
+      )}
+      {index < dropzoneChildrenLength - 1 && (
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: For clarity, please extract these conditions to vars above (`isFirstItem`, `isLastItem`). We should also have `isOnlyItem` and hide both buttons if the item is there by itself.

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 3]
```diff
@@ -15,29 +15,76 @@
...
+    return index;
+  }, [index]);
+
+  const dropzoneRecord = useMemo(
+    () => fromElement(getParentElementFromICEProps(modelId, fieldId, index)[0] as Element),
+    [fieldId, index, modelId]
+  );
+
+  const dropzoneChildrenLength = useMemo(
+    () => compileDropZone(dropzoneRecord.iceIds[0]).children.length,
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: If all you're getting from this is the number of children, it'd be best to not use the DOM and only use the ice registry to get this information.

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 4]
```diff
@@ -509,7 +509,7 @@ export function PreviewConcierge(props: any) {
             () => {
               issueDescriptorRequest({
                 site: siteId,
-                path,
+                path: path ?? models[parentModelId].craftercms.path,
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx:506]: In **each** case, looks like there are several expressions to get the path. Can we consolidate?

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 5]
```diff
@@ -48,23 +108,58 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
+          onTrash();
+          break;
+        }
+      }
+    };
+    document.addEventListener('keydown', onKeyDown);
+    return () => {
+      document.removeEventListener('keydown', onKeyDown);
+    };
+  }, [isFirstItem, isLastItem, onMoveDown, onMoveUp, onTrash]);
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: `onMoveDown`, `onMoveUp`, `onTrash` will trigger this effect every render.

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 6]
```diff
@@ -48,23 +108,58 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
     });
   };
   // endregion
+
+  useEffect(() => {
+    const onKeyDown = (e: KeyboardEvent) => {
+      switch (e.key) {
+        case 'ArrowUp': {
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx:123]: Add left arrow to this case

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 7]
```diff
@@ -48,23 +108,58 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
+          if (!isFirstItem) {
+            e.preventDefault();
+            onMoveUp();
+          }
+          break;
+        }
+        case 'ArrowDown': {
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx:130]: add right arrow to this case

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 8]
```diff
@@ -48,23 +108,58 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
+          if (!isLastItem) {
+            e.preventDefault();
+            onMoveDown();
+          }
+          break;
+        }
+        case 'Backspace': {
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx:136]: Add confirmation to this case

File: ui/guest/src/store/reducers/root.ts [hunk 9]
```diff
@@ -130,12 +130,13 @@ const reducer = createReducer(initialState, {
...
         const movableRecordId = iceRegistry.getMovableParentRecord(iceId);
-        if (notNullOrUndefined(movableRecordId)) {
+        const movableRecord = fromICEId(movableRecordId);
+        if (notNullOrUndefined(movableRecord)) {
           const draggable = getDraggable(movableRecordId);
           const highlight = getHoverData(
             // If (iceId == movableRecordId) the current record is already
             // the one to show the highlight on.
-            iceId === movableRecordId ? record.id : fromICEId(movableRecordId).id
+            iceId === movableRecordId ? record.id : movableRecord.id
```
Comment (rart) [inline ui/guest/src/store/reducers/root.ts]: Please explain these changes. What's wrong with checking movableRecordId without pulling the element record yet?

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 10]
```diff
@@ -15,29 +15,89 @@
...
+    fieldId: [fieldId],
+    index
+  } = record;
+
+  const elementIndex = useMemo(() => {
+    if (typeof index === 'string') {
+      return parseInt(index.substr(index.lastIndexOf('.') + 1), 10);
+    }
+    return index;
+  }, [index]);
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]:
```suggestion
  const elementIndex = useMemo(() => typeof index === 'string' ? parseInt(popPiece(index), 10) : index, [index]);
```

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 11]
```diff
@@ -15,29 +15,89 @@
...
+
+  const dropzoneRecord = useMemo(
+    () => fromElement(getParentElementFromICEProps(modelId, fieldId, index)[0] as Element),
+    [fieldId, index, modelId]
+  );
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: Likely not required

File: ui/guest/src/react/MoveModeZoneMenu.tsx (hunk 3 above)
Comment (rart) [reply ui/guest/src/react/MoveModeZoneMenu.tsx]: Try
```suggestion
    () => extractCollection(getCachedModel(modelId), fieldId, index).length,
```

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 12]
```diff
@@ -15,29 +15,89 @@
...
+
+  const onMoveDown = () => {
+    contentController.sortItem(
+      dropzoneRecord.modelId,
+      dropzoneRecord.fieldId[0],
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex}` : elementIndex,
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex + 1}` : elementIndex + 1
+    );
+    clearAndStartListening();
+  };
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: I think we should abstract this on the content controller: create `moveItemUp` & `moveItemDown` methods there and just invoke here. Method should take into account only items, index 0, last index and ignore updates in those cases.

File: ui/guest/src/store/reducers/root.ts (hunk 9 above)
Comment (aachavarria) [reply ui/guest/src/store/reducers/root.ts]: we can revert this change, i was trying to fix the issue when the items are moved and the registration is wrong

File: ui/app/src/modules/Preview/PreviewConcierge.tsx (hunk 4 above)
Comment (aachavarria) [reply ui/app/src/modules/Preview/PreviewConcierge.tsx:506]: almost all use the same, but there are a couple of exception

**Pull request merged**